include/sddf/block/blk_shared_queue.h

```diff
@@ -0,0 +1,342 @@
+                                        uint32_t *block_number,
+                                        uint16_t *count,
+                                        uint32_t *id)
+{
+    if (blk_req_queue_empty(queue_handle)) {
+        return -1;
+    }
+
+    *code = queue_handle->req_queue->buffers[queue_handle->req_queue->read_idx % queue_handle->req_queue->size].code;
+    *addr = queue_handle->req_queue->buffers[queue_handle->req_queue->read_idx % queue_handle->req_queue->size].addr;
+    *block_number = queue_handle->req_queue->buffers[queue_handle->req_queue->read_idx % queue_handle->req_queue->size].block_number;
+    *count = queue_handle->req_queue->buffers[queue_handle->req_queue->read_idx % queue_handle->req_queue->size].count;
+    *id = queue_handle->req_queue->buffers[queue_handle->req_queue->read_idx % queue_handle->req_queue->size].id;
+
+    THREAD_MEMORY_RELEASE();
+    queue_handle->req_queue->read_idx++;
+
+    return 0;
+}
+
+/**
+ * Dequeue an element from a response queue.
+ *
+ * @param queue queue handle containing response queue to dequeue from.
+ * @param status pointer to response status.
+ * @param addr pointer to encoded dma address of data.
+ * @param count pointer to number of blocks allocated for corresponding request
+ * @param success_count pointer to number of blocks successfully read/written
+ * @param id pointer to stoqueue request ID to idenfity which request this response is for.
+ * @return -1 when response queue is empty, 0 on success.
+ */
+static inline int blk_dequeue_resp(blk_queue_handle_t *queue_handle,
+                                        blk_response_status_t *status,
+                                        uintptr_t *addr,
+                                        uint16_t *count,
+                                        uint16_t *success_count,
+                                        uint32_t *id)
+{
+    if (blk_resp_queue_empty(queue_handle)) {
+        return -1;
+    }
+
+    *status = queue_handle->resp_queue->buffers[queue_handle->resp_queue->read_idx % queue_handle->resp_queue->size].status;
+    *addr = queue_handle->resp_queue->buffers[queue_handle->resp_queue->read_idx % queue_handle->resp_queue->size].addr;
+    *count = queue_handle->resp_queue->buffers[queue_handle->resp_queue->read_idx % queue_handle->resp_queue->size].count;
+    *success_count = queue_handle->resp_queue->buffers[queue_handle->resp_queue->read_idx % queue_handle->resp_queue->size].success_count;
+    *id = queue_handle->resp_queue->buffers[queue_handle->resp_queue->read_idx % queue_handle->resp_queue->size].id;
+
+    THREAD_MEMORY_RELEASE();
+    queue_handle->resp_queue->read_idx++;
+
+    return 0;
+}
+
+/**
+ * Set the plug of the request queue to true.
+ *
+ * @param queue_handle queue handle containing request queue to check for plug.
+*/
+static inline void blk_req_queue_plug(blk_queue_handle_t *queue_handle) {
+    queue_handle->req_queue->plugged = true;
```

Comment:
`plugged` seems to be a boolean that only gets set and checked by the users, and doesn't have anything to do with the implementation of the shared queue library despite the fact that the comment describes that it "prevents requests from being dequeued when plugged".

Reply:
The plugged is for optimisation where the driver is not meant to handle any subsequent requests when plugged. This functionality has to be communicated through some shared region + signalling protocol regardless, so it also make sense to put it in the shared library.